mods/bucket/init.lua

```diff
@@ -154,14 +153,17 @@ minetest.register_craftitem("bucket:bucket_empty", {
 			minetest.add_node(pointed_thing.under, {name="air"})
 
 			return ItemStack(giving_back)
-		elseif node_def then
-			-- Buckets will run a node's on_punch function if it is not liquid.
-			if node_def.on_punch then
-				node_def.on_punch(
-						pointed_thing.under,
-						minetest.get_node(pointed_thing.under),
-						user,
-						pointed_thing)
+		else
+			local node_def = minetest.registered_nodes[node.name]
+			if node_def then
+				-- Buckets will run a node's on_punch function if it is not liquid.
+				if node_def.on_punch then
+					node_def.on_punch(
+							pointed_thing.under,
+							minetest.get_node(pointed_thing.under),
+							user,
+							pointed_thing)
+				end
```

Comment:
You need to return an ItemStack here. If you don't return an ItemStack then the wielded item won't change. This is why it works one way, but not the other - the other code path neatly returns the empty bucket, but you never return the full bucket.